### ISSUE TITLE
Limitations of the JSON UI

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#2019]
I've spent a little time recently trying to build [a GTK-based front-end](https://gitlab.com/Screwtapello/kakoune-gtk) on top of Kakoune's JSON UI feature. It's not nearly complete, but I've already bumped into a few limitations of the JSON interface Kakoune exposes, and I figured I'd note them down here.

## No mouse event support

The JSON interface has a `keys()` message that tells Kakoune that keys were pressed, but it only supports keys that can be used with the `map` command—there's no way to tell Kakoune about mouse events, including scroll events and focus events.

## No clipboard integration support

One of the big selling points of the original Macintosh, and widely available in every GUI since, is the universal clipboard shared between all applications. It's always been difficult to make terminal applications interact with a GUI clipboard, so clipboard integration is a big potential benefit of a GUI-native Kakoune client.

Sadly, the JSON UI protocol does not provide any clipboard-related events in either direction, so a GUI-based Kakoune client is exactly as limited as the terminal UI.

## No mechanism for UI-specific events

Let's say my graphical UI wants to offer a "font" option, so people can switch what font is used. The obvious choice would be to add it to `ui_options` (that's what it's for!) but if the user changes that option, manually or via `kakrc`, there's no way for the UI to be notified and update the screen.

Another example: only the custom UI knows how to launch another instance of the custom UI (consider a front-end that does something like `ssh remotehost kak -ui json`), so in order for commands like `:new` to be possible, there needs to be a way for the server to tell the UI to open a new window.

## Atoms should include their character-cell width

The `draw()` message sends a terminal-sized block of text in the form of a sequence of `Atom` objects, where the character-cell width of all the Atoms on a line should add up to the total width. However, "character-cell width" is calculated from the data tables released by the Unicode consortium and different applications might have different versions of that data (and even if they use the same data, they still have to choose whether "ambiguous width" characters are 1 or 2 cells, see #1941), so when it comes to drawing Atoms to the screen, it doesn't so much matter what *I* think the width should be, or the Unicode database in *my* front-end thinks, I'll only get a sensible display if I follow what *Kakoune* thinks the width should be, and the only way for me to know for sure what Kakoune thinks the width should be is for Kakoune to include that data in each Atom.

## JSON UI protocol fundamentally tied to terminal UI

Beyond basic text display, the JSON UI supports "menu" and "info" messages. In theory, this would allow different front-ends to display that information in a way more "native" way, but in practice you're pretty much restricted to doing whatever the terminal UI does.

In most GUIs, a "menu" is a thing with a standard look-and-feel that can be navigated with the keyboard, clicked with the mouse, etc. Kakoune's menu system requires that the menu *not* accept keyboard focus (because key events must go to Kakoune) or mouse events (because the JSON protocol provides no way to signal that the user has selected something). Kakoune's menu-items can be structured (having multiple columns, etc) but the structure is presented as undifferentiated Atoms, so the only practical solution is to display them with the same monospaced font as everything else.

GUIs provide different widgets covering the different use-cases of Kakoune's info-box, from tool-tips ("inline" style) to modal dialogs ("modal" style), but a Kakoune UI can't really use those. For example, the "modal" style doesn't include any information about what the "buttons" of the modal dialog should be, so I can't display a real modal dialog, and even if I could I wouldn't be able to send the user's answer back to Kakoune because there's no "answer modal" message—the whole thing depends on the "modal" dialog being displayed inside the terminal window so ordinary terminal keypresses can be used to answer it.

The menu and info systems interact, too. Normally a menu is a vertical list, but when completion occurs in the prompt, it's displayed as a horizontal table so the info window can be displayed above it without overlapping. It's hard to imagine any way for those two things to appear at the same time without it being exactly the way it already works.

**Comment by lenormf** [comment]
It would also be nice to have a `notify-ui` method, to allow the UI to be notified of some events that happen internally (e.g. a new buffer was opened - that would allow constructing a buffer list in the frontend).

**Comment by mawww** [comment]
Hello,

thanks for starting this discussion, I'd be happy to improve the json-ui protocol to fix any issues alternates UIs have. Here are my thoughts on those:

* Mouse handling should be solved by making the mouse key syntax available. Currently we can print mouse key events, but not parse them. The UI should be able to send them (or, even better, change the key protocol to send a structured version of the key instead of a string).

* Changing ui_options is actually expected to send a notification to the UI, thats how the NCursesUI works. Looking at the code, it looks like I have been lazy, as JsonUI::set_ui_options does nothing. This needs to be fixed, and the documentation updated. There should be an rpc call for that.

* Starting another UI on the same session is solveable by having the UI define its own implementation of `:new`. Note that `:new` is currently an alias, depending on the situation it will alias to `x11-new` or `tmux-new-...` or `iterm-new-...`. The UI could just inject its own command, and alias `new` to it. Whats missing is a way for an UI to just say "run these commands". A special message from the UI to do that would be easy to add, and would open the way for UI to do more advanced interactions (like binding Kakoune commands to entries in a menu bar).

* Regarding menu and info being tied to terminal display, the goal was definitely for that not to be the case. What we lack is a way to support direct selection of menu items, there should be a `menu:select:<item number>` key to support that. Then mouse/keyboard support for selecting a menu entry would be doable by the UI by resolving the item number we want to select and sending a direct select key.

* Menus are voluntarily unstructured so that the UI can choose to display them as they want. The `style` parameter gives contextual information to the UI on what to do with them. `modal` is an info box because that was the simplest to implement, I agree it could be a dialog box with buttons, letting the UI take care of getting an answer, but that would be a pretty big change that is not really urgent for now. Regarding interaction between menu and info, the UIs are free to do as they want, the info style gives contextual information (for example, a `menuDoc` style means that info is related to the currently selected menu entry, and the UI should try to make that explicit).

Cheers,

Maxime.

**Comment by casimir** [comment]
I've spent a week-end hacking on a GUI myself. I've yet to implement menus but my experience has been good so far. On top of that I can see that @mawww has already addressed the issues on menu manipulation and mouse support. Lucky me.

About the clipboard integration I think this is already handled by scripting. I know a lot of users, including me, have mappings to interact with the system clipboard using xclipd/xsel/pbcopy/whatever. There is even a [wiki page](https://github.com/mawww/kakoune/wiki/Registers---Clipboard) that describe a solution that is not tied to the current client.

As for the atom width I'm not sure I understand the problem. This looks like a rendering issue. Every GUI toolkit comes with a way to get the real size of a given text if you need to tweak alignment. Also by using a monospaced font this is even easier to calculate by yourself.

I found a couple of annoying things on top of that:

### The status line is both a command buffer and a message display

I'm trying to implement a command palette so I need to differentiate a command edition from a message display. My strategy as of now is to consider a line beginning with `:` as a command edition. But it's not always correct.

I also plan to display errors differently than messages, in a toast notification for example. In the current state the only way I can imagine is to parse the message's face and compare it to the current `Error` face. That doesn't look robust at all...

### No way to capture a command output easily

Let's say we want to get an option's value or the buffer list from the GUI. It's basically capturing an `echo` output. 

We could create a scratch buffer to get the output but that implies a lot of setup to handle synchronicity and avoid clashing with other command calls (a GUI generally implies a lot of parallelism).

Another *hacky* way is to abuse the status line. By using a "smart" prefix mapping command result is feasible, a prefix being something like `>cmd_id>cmd result`. I got some results with this method but it's pretty ugly.

The right way would be to have a new RPC call to execute the command and respond with the result and output instead of "echoing" it in the status line.
Since the RPC implementation is synchronous in our case it should be easy to implement and even easier to use.

### No `id` attribute in RPC responses

This one is more for correctness than usability since the implementation is synchronous. The [JSON-RPC standard](http://www.jsonrpc.org/specification#response_object) specifies that an `id` attribute must be set in requests objects and their responses but kakoune omits this attribute. This prevents the usage of most JSON-RPC libraries.
I think it would be a quick win to add this.


That's all for me. Sorry for the long post.

**Comment by Screwtapello** [comment]
> As for the atom width I'm not sure I understand the problem.

I can see I was a bit rambly, let me try again.

- Kakoune has various features that require text on different lines to be vertically aligned: the `&` command, filenames displayed in the completion menu, the `column` highlighter, etc.
- Because Kakoune is Unicode-aware, it measures text with the `wcwidth()` libc function to find out how many columns it should take, and the `Atom`s it creates are padded and truncated accordingly.
- In a graphical front-end, we can choose how wide to draw a piece of text, but we don't want any old width—we want to match Kakoune's expected width so that things will line up properly.
    - We can just let the font's native spacing and kerning determine the width, which may or may not line up with what Kakoune expects (probably not for a proportional font, probably does for a monospaced font, unless it's a monospaced font [designed for East Asian text](https://github.com/mawww/kakoune/issues/1941)).
    - We can call `wcwidth()` ourselves to try and reconstruct the information Kakoune used, which works if the front-end and back-end are using the same version of libc with the same locale. However, that's not guaranteed: the server might be running on a different host via SSH, or the client might be written in a language with [its](https://github.com/mycoboco/wcwidth.js) [own](https://github.com/jquast/wcwidth) [character](https://github.com/lucy/wcwidth.rs) [width](https://github.com/mattn/go-runewidth) [implementation](https://github.com/janlelis/unicode-display_width) that may be based on a different version of the Unicode standard and hence disagree with libc.
    - Or, Kakoune could just include the expected width as part of the `Atom` structure and eliminate all guessing.

In short, to ensure consistency we should measure text once, and then use that measurement everywhere. Measuring multiple times with different rules (font metrics, different copies of different versions of Unicode data) can only lead to heartache.

> About the clipboard integration I think this is already handled by scripting.

While I agree that clipboard interaction is *possible*, I can't quite bring myself to call it *integrated*.

- As a person used to X11, I might expect that dragging across some text would "select" it so I can middle-click-paste it into another application. This will probably never work with Kakoune because of its novel movement-is-selection model.
- As a person used to GUIs, I might expect that I can select some text and then press a toolbar button, or hit Ctrl-C to copy it. This can't be done right now because there's no way for the UI to reach into the JSON client and examine the current selection. Some kind of "evaluate-commands" JSON message would help here.
- As a person used to GUIs, I might expect that I can copy some text in another application and then press a toolbar button or hit Ctrl-V to paste it. This can be done in the same way it works in the terminal ("squirt text at Kakoune and see what happens"), but one of the nice things about a GUI is that we *can* distinguish a paste from really fast typing. I'd love to be able to do something like `map global insert <paste:start> <a-;>\i` and `map global insert <paste:end> <esc>`.
- As a person used to terminal Kakoune where 'y' copies to a default register, I might expect a GUI Kakoune to have 'y' copy to the GUI "default register", the clipboard. This can't be done right now because there's no way for the UI to observe changes to registers.
- As a person used to terminal Kakoune where 'p' pastes from a default register, I might expect a GUI Kakoune to have 'p' paste from the GUI "default register", the clipboard. This can't be done right now because there's no way for the UI to interrupt reads from registers, and no way to tell whether Kakoune will interpret a given "p" press as a paste, or typing, or selecting the "p" register, or a user-mode command...

A JSON UI already needs to bridge between the Kakoune world and the front-end resources like text-rendering and input-handling. Having to use other tools to bridge between the Kakoune world and equally important resources like the clipboard feels clunky and redundant to me.

> No id attribute in RPC responses

Kakoune's API doesn't really have responses, just requests in both directions that don't expect replies. Looking at the spec, these are termed [Notifications](http://www.jsonrpc.org/specification#notification), and it says: "A Notification is a Request object without an "id" member." So I think Kakoune is already doing the right thing here, but perhaps the documentation could be made clearer that there are no responses involved.

**Comment by Screwtapello** [comment]
Some more issues I've encountered:

- no way to communicate modifiers with mouse-clicks, so the text UI's "control-click to add a new cursor" feature can't be implemented
- When sending a bogus JSON-RPC message, Kakoune prints the resulting error message to `stderr`, but does not print a newline afterward, so the errors all stack up and are hard to read.

**Comment by vbauerster** [comment]
Why don't collaborate with [Oni](https://github.com/onivim/oni/issues/2132), to support kakoune?
I hope this isn't very utopian idea )

**Comment by Screwtapello** [comment]
Looking at the [Oni Editor Interface](https://onivim.github.io/oni-api/interfaces/editor.html) linked from that issue, the very first property on the `Editor` object is `activeBuffer`, but Kakoune's UI protocol doesn't tell you anything about the active buffer, or what buffers are open, or when the user switches from one buffer to another, so I don't think it's really going to be practical to add Kakoune support to Oni unless Kakoune's JSON-UI gets dramatically more complex.

I imagine it would be a lot easier to make NeoVim speak Kakoune's much more limited UI protocol, but of course then you'd get to pick from all the half-finished alternate Kakoune UIs. :)
